### DISCUSSION TITLE
pyslim.recapitate( ) Errors

### DISCUSSION THREAD

**Discussion by KNLawrence01** [tskit-dev/pyslim#268] [Q&A] [answered]
Hey all, 

So I've been working further on my last problem and have ensured that I'm using v0.700 of pyslim, but I am still coming across some problems. For the recapitation of the tree I have now described demography, but am still having no luck with pyslim.recapitate( ). 

import msprime, pyslim
import tskit

ts = tskit.load("MS123sweep.trees")

demography = msprime.Demography.from_tree_sequence(ts, initial_size=20000)
demography.add_population_split(time =400 , derived = ["pop_3"], ancestral = "pop_2")
demography.add_population_split(time =4000 , derived = ["pop_2", "pop_1"], ancestral = "pop_0")
print(demography.debug())

rts = pyslim.recapitate(ts, demography=demography, recombination_rate= 1.14e-8, random_seed=1)

recap.dump("MS123sweepRecap.trees")

After Using this script, I get the debug of the demography and this error:

/opt/miniconda3/lib/python3.8/site-packages/msprime/ancestry.py:831: TimeUnitsMismatchWarning: The initial_state has time_units=unknown but time is measured in generations in msprime. This may lead to significant discrepancies between the timescales. If you wish to suppress this warning, you can use, e.g., warnings.simplefilter('ignore', msprime.TimeUnitsMismatchWarning)
  warnings.warn(message, TimeUnitsMismatchWarning)
/opt/miniconda3/lib/python3.8/site-packages/msprime/demography.py:1142: IncompletePopulationMetadataWarning: The metadata schema does not have a 'name' property; population names will not be recorded in the output tree sequence
  warnings.warn(
/opt/miniconda3/lib/python3.8/site-packages/msprime/demography.py:1149: IncompletePopulationMetadataWarning: The metadata schema does not have a 'description' property; population descriptions will not be recorded in the output tree sequence
  warnings.warn(
Traceback (most recent call last):
  File "recapSLiMmsprime.txt", line 13, in <module>
    rts = pyslim.recapitate(ts, demography=demography, recombination_rate= 1.14e-8, random_seed=1)
  File "/opt/miniconda3/lib/python3.8/site-packages/pyslim/methods.py", line 78, in recapitate
    recap = msprime.sim_ancestry(
  File "/opt/miniconda3/lib/python3.8/site-packages/msprime/ancestry.py", line 1207, in sim_ancestry
    sim = _parse_sim_ancestry(
  File "/opt/miniconda3/lib/python3.8/site-packages/msprime/ancestry.py", line 1010, in _parse_sim_ancestry
    return Simulator(
  File "/opt/miniconda3/lib/python3.8/site-packages/msprime/ancestry.py", line 1318, in __init__
    super().__init__(
msprime._msprime.InputError: Input error in initialise: Attempt to sample a lineage from an inactive population

Thanks!

**Comment by petrelharp** [comment, marked as answer]
*(posted this to the SLiM list also)*

Let's see: there's some kind of mismatch between where the
lineages are in the tree sequence and the demography. I can't be sure
without knowing some more things about the SLiM recipe, but my first
guess is that your SLiM simulation ran for 400 generations, and
everyone there is in population 3?  This would create the error that
you're seeing because add_subpopulation_split(time=400, derived,
ancestral) implies that more recently than 400 time units ago, the
derived pops are "inactive" (ie didn't exist), so that all lineages
should shift from the derived populations to the ancestral one. This
means it's an error if there are any lineages in the derived pops
before time=400. The way that time intervals work in msprime, however,
it's an error if there are lineages in the derived pops with time >=
400; and if your SLiM simulation ran for 400 generations, then that
will be the time of the lineages you're trying to recapitate. The fix
is to change 400 to 401 (or any number larger than 400).

If that doesn't fix it, you can check where and when the lineages
needing recapitation are by something like:
```
for p, t in set([ts.node(r).population, ts.node(r).time for t in
ts.trees for r in t.roots]):
    print(f"pop {p}, time {t}")
```
and then printing your demography to make sure things match up.
